{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2019a3de-4963-489e-b5e0-6fa36ce66aac",
   "metadata": {},
   "source": [
    "## DASH BOARD CONSTRUCTION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dff117f0-66ad-40b7-ab55-3c690d0566e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import dash\n",
    "from dash import dcc\n",
    "from dash import html\n",
    "from dash.dependencies import Input, Output\n",
    "from sqlalchemy import create_engine\n",
    "from dash import dash_table\n",
    "import psycopg2\n",
    "import plotly.figure_factory as ff\n",
    "import plotly.express as px \n",
    "import os \n",
    "POSTGRES_PASSWORD = os.getenv('POSTGRES_PASSWORD')\n",
    "import pymongo\n",
    "import dash_bootstrap_components as dbc\n",
    "from bson.json_util import loads, dumps\n",
    "mongo_username = os.getenv('MONGO_INITDB_ROOT_USERNAME')\n",
    "mongo_password = os.getenv('MONGO_INITDB_ROOT_PASSWORD')\n",
    "mongo_init_db = os.getenv('MONGO_INITDB_DATABASE')\n",
    "import matplotlib.pyplot as plt\n",
    "import string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b1ff84fe-7601-4f3d-b9b4-f0d143a463a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## building the engine for the database\n",
    "engine = create_engine(\"postgresql+psycopg2://{user}:{pw}@{service}:{port}/{db}\".format(\n",
    "    user = 'postgres', \n",
    "    pw = POSTGRES_PASSWORD, \n",
    "    service= \"postgres\",\n",
    "    port=5432, \n",
    "    db = 'movie', \n",
    "))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "913fdc7f-b559-445a-8106-34315afeff9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "external_stylesheets = ['https://codepen.io/chriddyp/pen/bWLwgP.css']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "75a4502e-4631-42f0-a269-11124a4e9542",
   "metadata": {},
   "outputs": [],
   "source": [
    "mymarkdown = '''\n",
    "    *Acquire API key from TMDb: [TMDb API](https://developer.themoviedb.org/docs)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "26618e48-eda8-4512-9819-3416447fd124",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Animation', 'Comedy', 'Family', 'Music', 'Fantasy', 'Drama',\n",
       "       'History', 'Science Fiction', 'Action', 'Thriller', 'Horror',\n",
       "       'Mystery', 'Adventure', 'Crime', 'War', 'Romance', 'Western',\n",
       "       'TV Movie', 'Documentary'], dtype=object)"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_df = pd.read_csv(\"genres.csv\")\n",
    "genre_list = genre_df['genre'].dropna().unique()\n",
    "genre_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "ae200f2a-f21d-4ede-b731-6bfc736b4ef8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['en', 'hi', 'mn', 'ko', 'ja', 'id', 'es', 'fr', 'it', 'uk', 'zh',\n",
       "       'th', 'ar', 'el', 'nl', 'tr', 'te', 'de', 'pt', 'fi', 'ml', 'ru',\n",
       "       'is', 'pl', 'ta', 'cn', 'sv', 'no', 'da', 'tl', 'yo', 'bn', 'sr',\n",
       "       'xh', 'cs', 'mk', 'kk', 'gl', 'fa', 'ga', 'kn', 'lv', 'ca', 'pa',\n",
       "       'sh', 'km', 'nb', 'ms', 'lt', 'hu', 'ro', 'he', 'eu', 'bg'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "movie_df = pd.read_csv(\"movies.csv\",lineterminator='\\n')\n",
    "language_list = movie_df['language'].dropna().unique()\n",
    "language_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "ec6f38d5-9bea-4ec7-bfc5-0c3d048382bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "iso_639_choices = [('ab', 'Abkhaz'),('aa', 'Afar'),('af', 'Afrikaans'),('ak', 'Akan'),('sq', 'Albanian'),('am', 'Amharic'),('ar', 'Arabic'),('an', 'Aragonese'),('hy', 'Armenian'),('as', 'Assamese'),('av', 'Avaric'),('ae', 'Avestan'),('ay', 'Aymara'),('az', 'Azerbaijani'),('bm', 'Bambara'),('ba', 'Bashkir'),('eu', 'Basque'),('be', 'Belarusian'),('bn', 'Bengali'),('bh', 'Bihari'),('bi', 'Bislama'),('bs', 'Bosnian'),('br', 'Breton'),('bg', 'Bulgarian'),('my', 'Burmese'),('ca', 'Catalan; Valencian'),('ch', 'Chamorro'),('ce', 'Chechen'),('ny', 'Chichewa; Chewa; Nyanja'),('zh', 'Chinese'),('cv', 'Chuvash'),('kw', 'Cornish'),('co', 'Corsican'),('cr', 'Cree'),('hr', 'Croatian'),('cs', 'Czech'),('da', 'Danish'),('dv', 'Divehi; Maldivian;'),('nl', 'Dutch'),('dz', 'Dzongkha'),('en', 'English'),('eo', 'Esperanto'),('et', 'Estonian'),('ee', 'Ewe'),('fo', 'Faroese'),('fj', 'Fijian'),('fi', 'Finnish'),('fr', 'French'),('ff', 'Fula'),('gl', 'Galician'),('ka', 'Georgian'),('de', 'German'),('el', 'Greek, Modern'),('gn', 'Guaraní'),('gu', 'Gujarati'),('ht', 'Haitian'),('ha', 'Hausa'),('he', 'Hebrew (modern)'),('hz', 'Herero'),('hi', 'Hindi'),('ho', 'Hiri Motu'),('hu', 'Hungarian'),('ia', 'Interlingua'),('id', 'Indonesian'),('ie', 'Interlingue'),('ga', 'Irish'),('ig', 'Igbo'),('ik', 'Inupiaq'),('io', 'Ido'),('is', 'Icelandic'),('it', 'Italian'),('iu', 'Inuktitut'),('ja', 'Japanese'),('jv', 'Javanese'),('kl', 'Kalaallisut'),('kn', 'Kannada'),('kr', 'Kanuri'),('ks', 'Kashmiri'),('kk', 'Kazakh'),('km', 'Khmer'),('ki', 'Kikuyu, Gikuyu'),('rw', 'Kinyarwanda'),('ky', 'Kirghiz, Kyrgyz'),('kv', 'Komi'),('kg', 'Kongo'),('ko', 'Korean'),('ku', 'Kurdish'),('kj', 'Kwanyama, Kuanyama'),('la', 'Latin'),('lb', 'Luxembourgish'),('lg', 'Luganda'),('li', 'Limburgish'),('ln', 'Lingala'),('lo', 'Lao'),('lt', 'Lithuanian'),('lu', 'Luba-Katanga'),('lv', 'Latvian'),('gv', 'Manx'),('mk', 'Macedonian'),('mg', 'Malagasy'),('ms', 'Malay'),('ml', 'Malayalam'),('mt', 'Maltese'),('mi', 'Māori'),('mr', 'Marathi (Marāṭhī)'),('mh', 'Marshallese'),('mn', 'Mongolian'),('na', 'Nauru'),('nv', 'Navajo, Navaho'),('nb', 'Norwegian Bokmål'),('nd', 'North Ndebele'),('ne', 'Nepali'),('ng', 'Ndonga'),('nn', 'Norwegian Nynorsk'),('no', 'Norwegian'),('ii', 'Nuosu'),('nr', 'South Ndebele'),('oc', 'Occitan'),('oj', 'Ojibwe, Ojibwa'),('cu', 'Old Church Slavonic'),('om', 'Oromo'),('or', 'Oriya'),('os', 'Ossetian, Ossetic'),('pa', 'Panjabi, Punjabi'),('pi', 'Pāli'),('fa', 'Persian'),('pl', 'Polish'),('ps', 'Pashto, Pushto'),('pt', 'Portuguese'),('qu', 'Quechua'),('rm', 'Romansh'),('rn', 'Kirundi'),('ro', 'Romanian, Moldavan'),('ru', 'Russian'),('sa', 'Sanskrit (Saṁskṛta)'),('sc', 'Sardinian'),('sd', 'Sindhi'),('se', 'Northern Sami'),('sm', 'Samoan'),('sg', 'Sango'),('sr', 'Serbian'),('gd', 'Scottish Gaelic'),('sn', 'Shona'),('si', 'Sinhala, Sinhalese'),('sk', 'Slovak'),('sl', 'Slovene'),('so', 'Somali'),('st', 'Southern Sotho'),('es', 'Spanish; Castilian'),('su', 'Sundanese'),('sw', 'Swahili'),('ss', 'Swati'),('sv', 'Swedish'),('ta', 'Tamil'),('te', 'Telugu'),('tg', 'Tajik'),('th', 'Thai'),('ti', 'Tigrinya'),('bo', 'Tibetan'),('tk', 'Turkmen'),('tl', 'Tagalog'),('tn', 'Tswana'),('to', 'Tonga'),('tr', 'Turkish'),('ts', 'Tsonga'),('tt', 'Tatar'),('tw', 'Twi'),('ty', 'Tahitian'),('ug', 'Uighur, Uyghur'),('uk', 'Ukrainian'),('ur', 'Urdu'),('uz', 'Uzbek'),('ve', 'Venda'),('vi', 'Vietnamese'),('vo', 'Volapük'),('wa', 'Walloon'),('cy', 'Welsh'),('wo', 'Wolof'),('fy', 'Western Frisian'),('xh', 'Xhosa'),('yi', 'Yiddish'),('yo', 'Yoruba'),('za', 'Zhuang, Chuang'),('zu', 'Zulu'),('zh',\"Chinese\")]\n",
    "\n",
    "code_list = [i[0] for i in iso_639_choices]\n",
    "value_list = [i[1] for i in iso_639_choices]\n",
    "lan_df = pd.DataFrame(data={\"code\":code_list, \"value\":value_list})\n",
    "lan_df = lan_df[lan_df['code'].isin(language_list) ]\n",
    "languagelist = [{\"label\":x,'value':y} for x,y in zip(lan_df['value'], lan_df[\"code\"])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "id": "3e9b2bf1-ab85-4105-b0c5-e4d4c10ffe72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"650\"\n",
       "            src=\"http://0.0.0.0:8050/\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0xffff1b122d50>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2023-12-10 03:36:16,376] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "psycopg2.errors.SyntaxError: syntax error at or near \"s\"\n",
      "LINE 4:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                 ^\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 1455, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 869, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 867, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 852, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/dash.py\", line 1310, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_282/2308065810.py\", line 125, in display_button_property\n",
      "    movie = pd.read_sql_query(myquery, con = engine)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 469, in read_sql_query\n",
      "    return pandas_sql.read_query(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1738, in read_query\n",
      "    result = self.execute(sql, params)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1562, in execute\n",
      "    return self.con.exec_driver_sql(sql, *args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1778, in exec_driver_sql\n",
      "    ret = self._execute_context(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1848, in _execute_context\n",
      "    return self._exec_single_context(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1988, in _exec_single_context\n",
      "    self._handle_dbapi_exception(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 2343, in _handle_dbapi_exception\n",
      "    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near \"s\"\n",
      "LINE 4:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                 ^\n",
      "\n",
      "[SQL: \n",
      "    SELECT  *\n",
      "    FROM movies m\n",
      "    WHERE title = 'Five Nights at Freddy's'\n",
      "    ]\n",
      "(Background on this error at: https://sqlalche.me/e/20/f405)\n",
      "[2023-12-10 03:36:16,388] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "psycopg2.errors.SyntaxError: syntax error at or near \"s\"\n",
      "LINE 12:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                  ^\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 1455, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 869, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 867, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 852, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/dash.py\", line 1310, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_282/2308065810.py\", line 147, in display_movie_table\n",
      "    movie = pd.read_sql_query(myquery, con = engine)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 469, in read_sql_query\n",
      "    return pandas_sql.read_query(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1738, in read_query\n",
      "    result = self.execute(sql, params)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1562, in execute\n",
      "    return self.con.exec_driver_sql(sql, *args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1778, in exec_driver_sql\n",
      "    ret = self._execute_context(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1848, in _execute_context\n",
      "    return self._exec_single_context(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1988, in _exec_single_context\n",
      "    self._handle_dbapi_exception(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 2343, in _handle_dbapi_exception\n",
      "    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near \"s\"\n",
      "LINE 12:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                  ^\n",
      "\n",
      "[SQL: \n",
      "    SELECT  m.title,\n",
      "            l.value,\n",
      "            m.runtime,\n",
      "            m.score,\n",
      "            g.genre\n",
      "    FROM movies m\n",
      "    INNER JOIN language l\n",
      "        ON m.language = l.code\n",
      "    INNER JOIN genres g\n",
      "        ON m.id = g.id\n",
      "    WHERE title = 'Five Nights at Freddy's'\n",
      "    ]\n",
      "(Background on this error at: https://sqlalche.me/e/20/f405)\n",
      "[2023-12-10 03:36:16,437] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "psycopg2.errors.SyntaxError: syntax error at or near \"s\"\n",
      "LINE 4:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                 ^\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 1455, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 869, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 867, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 852, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/dash.py\", line 1310, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_282/2308065810.py\", line 204, in put_map\n",
      "    movie = pd.read_sql_query(myquery, con = engine)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 469, in read_sql_query\n",
      "    return pandas_sql.read_query(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1738, in read_query\n",
      "    result = self.execute(sql, params)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1562, in execute\n",
      "    return self.con.exec_driver_sql(sql, *args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1778, in exec_driver_sql\n",
      "    ret = self._execute_context(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1848, in _execute_context\n",
      "    return self._exec_single_context(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1988, in _exec_single_context\n",
      "    self._handle_dbapi_exception(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 2343, in _handle_dbapi_exception\n",
      "    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near \"s\"\n",
      "LINE 4:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                 ^\n",
      "\n",
      "[SQL: \n",
      "    SELECT  *\n",
      "    FROM movies m\n",
      "    WHERE title = 'Five Nights at Freddy's'\n",
      "    ]\n",
      "(Background on this error at: https://sqlalche.me/e/20/f405)\n",
      "[2023-12-10 03:36:16,439] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "psycopg2.errors.SyntaxError: syntax error at or near \"s\"\n",
      "LINE 4:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                 ^\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 1455, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 869, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 867, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 852, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/dash.py\", line 1310, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_282/2308065810.py\", line 168, in display_decription\n",
      "    movie = pd.read_sql_query(myquery, con = engine)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 469, in read_sql_query\n",
      "    return pandas_sql.read_query(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1738, in read_query\n",
      "    result = self.execute(sql, params)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1562, in execute\n",
      "    return self.con.exec_driver_sql(sql, *args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1778, in exec_driver_sql\n",
      "    ret = self._execute_context(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1848, in _execute_context\n",
      "    return self._exec_single_context(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1988, in _exec_single_context\n",
      "    self._handle_dbapi_exception(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 2343, in _handle_dbapi_exception\n",
      "    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near \"s\"\n",
      "LINE 4:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                 ^\n",
      "\n",
      "[SQL: \n",
      "    SELECT *\n",
      "    FROM movies m\n",
      "    WHERE title = 'Five Nights at Freddy's'\n",
      "    ]\n",
      "(Background on this error at: https://sqlalche.me/e/20/f405)\n",
      "[2023-12-10 03:36:16,439] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "psycopg2.errors.SyntaxError: syntax error at or near \"s\"\n",
      "LINE 12:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                  ^\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 1455, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 869, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 867, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 852, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/dash.py\", line 1310, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_282/2308065810.py\", line 240, in cast_image\n",
      "    cast = pd.read_sql_query(myquery, con = engine)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 469, in read_sql_query\n",
      "    return pandas_sql.read_query(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1738, in read_query\n",
      "    result = self.execute(sql, params)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1562, in execute\n",
      "    return self.con.exec_driver_sql(sql, *args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1778, in exec_driver_sql\n",
      "    ret = self._execute_context(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1848, in _execute_context\n",
      "    return self._exec_single_context(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1988, in _exec_single_context\n",
      "    self._handle_dbapi_exception(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 2343, in _handle_dbapi_exception\n",
      "    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near \"s\"\n",
      "LINE 12:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                  ^\n",
      "\n",
      "[SQL: \n",
      "    SELECT c.character,\n",
      "        cm.name,\n",
      "        cm.popularity,\n",
      "        cm.profile_path,\n",
      "        cm.known_for_department\n",
      "    FROM movies m\n",
      "    INNER JOIN movie_cast c\n",
      "        ON m.id = c.movie_id\n",
      "    INNER JOIN cast_members cm\n",
      "        ON c.cast_member_id = cm.cast_member_id\n",
      "    WHERE title = 'Five Nights at Freddy's'\n",
      "    ORDER BY cm.popularity DESC\n",
      "    LIMIT 15\n",
      "    ]\n",
      "(Background on this error at: https://sqlalche.me/e/20/f405)\n",
      "[2023-12-10 03:36:16,446] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "psycopg2.errors.SyntaxError: syntax error at or near \"s\"\n",
      "LINE 4:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                 ^\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 1455, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 869, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 867, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/flask/app.py\", line 852, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/dash.py\", line 1310, in dispatch\n",
      "    ctx.run(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/dash/_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_282/2308065810.py\", line 269, in display_comments\n",
      "    movie = pd.read_sql_query(myquery, con = engine)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 469, in read_sql_query\n",
      "    return pandas_sql.read_query(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1738, in read_query\n",
      "    result = self.execute(sql, params)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/pandas/io/sql.py\", line 1562, in execute\n",
      "    return self.con.exec_driver_sql(sql, *args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1778, in exec_driver_sql\n",
      "    ret = self._execute_context(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1848, in _execute_context\n",
      "    return self._exec_single_context(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1988, in _exec_single_context\n",
      "    self._handle_dbapi_exception(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 2343, in _handle_dbapi_exception\n",
      "    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py\", line 1969, in _exec_single_context\n",
      "    self.dialect.do_execute(\n",
      "  File \"/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py\", line 922, in do_execute\n",
      "    cursor.execute(statement, parameters)\n",
      "sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near \"s\"\n",
      "LINE 4:     WHERE title = 'Five Nights at Freddy's'\n",
      "                                                 ^\n",
      "\n",
      "[SQL: \n",
      "    SELECT  *\n",
      "    FROM movies m\n",
      "    WHERE title = 'Five Nights at Freddy's'\n",
      "    ]\n",
      "(Background on this error at: https://sqlalche.me/e/20/f405)\n"
     ]
    }
   ],
   "source": [
    "from dash import Dash, dcc, html, Input, Output, State, callback, ALL, MATCH\n",
    "import plotly.graph_objs as go\n",
    "app = dash.Dash(__name__, external_stylesheets=external_stylesheets)\n",
    "app.layout = html.Div(\n",
    "    [\n",
    "        # stuff on top of the dashboard\n",
    "        html.H2(\"Movie Information Dashboard\",style={'textAlign': 'center',\"font-weight\": \"bold\"}), \n",
    "        html.H4('The Dashboard that is built for the movie lovers',style={'textAlign': 'center'}),\n",
    "        # side bar \n",
    "        html.Div([\n",
    "            # recommend movie display panel (for the drop down)\n",
    "            html.Div([\n",
    "                 dcc.Dropdown(id = 'recommed_dropdown',placeholder=\"Select a movie\",value = \"Puss in Boots: The Last Wish\"),\n",
    "            ],id = \"movie_display_panel\",style={\"border\":\"2px\",'height':'36px','width':'74%','float':'right',\"border\":\"2px black solid\", 'margin':\"5px\"}),\n",
    "            dcc.Markdown(\"Press 'Generate' button down below, then use drop down menu to select movies based on your preferences\", style={'textAlign': 'center','width':'24%', 'float':'left',\"margin_top\":\"10px\",\"margin_bottom\":\"10px\",\"font-weight\": \"bold\"}),  \n",
    "        ],style = { 'height': '100px', 'width':'100%',\"border\":\"2px black solid\", 'margin-right':\"5px\", 'margin-bottom':'5px'}),\n",
    "\n",
    "        # side bar for displaying the user preferences\n",
    "        html.Div([\n",
    "            dcc.Markdown(\"Please select your preferences\",id = \"display_preferences\",style={\"font-weight\": \"bold\"}),\n",
    "            # movie release year selection: \n",
    "            dcc.Markdown(\"Movie Release Year: \"),\n",
    "            dcc.Slider(1980, 2023, 1, id = \"release_year\", value = 2023,marks=None,\n",
    "    tooltip={\"placement\": \"bottom\", \"always_visible\": True}),\n",
    "            # movie genre selection:\n",
    "            dcc.Markdown(\"Genre:\"),\n",
    "            dcc.Dropdown(id = 'genre_list',\n",
    "                        options = genre_list,\n",
    "                         value = \"Action\"),\n",
    "            # movie minimum runtime selection:\n",
    "            dcc.Markdown(\"Minimum Movie Run Time (min): \"),\n",
    "            dcc.Slider(20, 150, 5, id=\"runtime\", marks=None, value = 60,\n",
    "    tooltip={\"placement\": \"bottom\", \"always_visible\": True}),\n",
    "            # movie language selection: \n",
    "            dcc.Markdown(\"Language: \"),\n",
    "            dcc.Dropdown(id = 'lanList',\n",
    "                        options = languagelist,\n",
    "                        value = \"en\"),\n",
    "            # movie minimum user rating selection: \n",
    "            dcc.Markdown(\"Minimum User Rating: \"),\n",
    "            dcc.Slider(0,10,1, marks=None, id = 'rating', value = 5,\n",
    "    tooltip={\"placement\": \"bottom\", \"always_visible\": True}),\n",
    "            dcc.Markdown(mymarkdown),\n",
    "            \n",
    "        ],style = {'width':'24%', 'float':'left'}),\n",
    "        \n",
    "        # main bar\n",
    "        html.Div([\n",
    "            dcc.Tabs([\n",
    "               dcc.Tab(label = 'General Information', children = [\n",
    "                    # stuff for general information tab goes here\n",
    "                    html.Div([html.Img(id = 'movie_poster', alt='image',style={'height':'100%', 'width':'100%'})], style = {'width': '24%', 'float':'left','margin':'10px'}),\n",
    "                    html.Div([\n",
    "                            dcc.Graph(id = 'movie_table', style = {\"height\": '250px',}),\n",
    "                            dcc.Markdown(id = 'description', style = {'height':'300px',\"font-weight\": \"bold\",\"margin-top\":'10px'})\n",
    "                             ], style = {'width': '74%', 'float':'right'})\n",
    "                ]),\n",
    "                dcc.Tab(label = 'Cast Members', children = [\n",
    "                    # stuff for cast members goes here\n",
    "                           html.Div([],id = 'cast_image', style= {\"display\": \"inline-block\",})\n",
    "                   ]),\n",
    "                dcc.Tab(label = 'Movie Release Information', children = [\n",
    "                   # stuff for movie release goes here\n",
    "                     html.Div([\n",
    "                        html.H2 (\"Movie Release Date Information\",style={'textAlign': 'center',\"font-weight\": \"bold\"}),\n",
    "                        dcc.Graph(id= 'release_map')\n",
    "                    ],id = 'release_map_container')\n",
    "               ]),\n",
    "                dcc.Tab(label = 'Movie Reviews', children = [\n",
    "                   # stuff for reviews tab goes here\n",
    "                    html.Div([   \n",
    "                    ],id=\"display_panel\")\n",
    "                   \n",
    "               ]),\n",
    "            ])\n",
    "            \n",
    "        ],style = {'width':'74%', 'float':'right'})   \n",
    "        \n",
    "    ]\n",
    ")\n",
    "\n",
    "#call back for the display panel drop down\n",
    "\n",
    "@app.callback(\n",
    "    Output('recommed_dropdown','options'),\n",
    "    [Input('release_year', 'value'),\n",
    "    Input('lanList', 'value'),\n",
    "    Input('genre_list','value'),\n",
    "    Input('runtime', 'value'),\n",
    "    Input('rating', 'value')]\n",
    ")\n",
    "def update_output(year, lan, genre, runtime, rating):\n",
    "    current_buttons = []\n",
    "    myquery = f'''\n",
    "    SELECT  *\n",
    "    FROM movies m\n",
    "        INNER JOIN genres g\n",
    "        ON m.id = g.id\n",
    "    WHERE year = {year} \n",
    "        AND language = '{lan}'\n",
    "        AND genre = '{genre}'\n",
    "        AND runtime >= {runtime}\n",
    "        AND score >= {rating}\n",
    "    ORDER BY score DESC\n",
    "    '''\n",
    "    movie = pd.read_sql_query(myquery, con = engine)#.T.reset_index()\n",
    "    return movie['title']\n",
    "\n",
    "\n",
    "# call back for the poster image\n",
    "@app.callback(\n",
    "    Output('movie_poster', \"src\"),\n",
    "    Input('recommed_dropdown', 'value'),\n",
    ")\n",
    "def display_button_property(value):\n",
    "    myquery = f'''\n",
    "    SELECT  *\n",
    "    FROM movies m\n",
    "    WHERE title = '{value}'\n",
    "    '''\n",
    "    movie = pd.read_sql_query(myquery, con = engine)\n",
    "    url = movie['poster_image'][0]\n",
    "    return url\n",
    "    \n",
    "@app.callback(\n",
    "    Output('movie_table', \"figure\"),\n",
    "    Input('recommed_dropdown', 'value')\n",
    ")\n",
    "def display_movie_table(value):\n",
    "    myquery = f'''\n",
    "    SELECT  m.title,\n",
    "            l.value,\n",
    "            m.runtime,\n",
    "            m.score,\n",
    "            g.genre\n",
    "    FROM movies m\n",
    "    INNER JOIN language l\n",
    "        ON m.language = l.code\n",
    "    INNER JOIN genres g\n",
    "        ON m.id = g.id\n",
    "    WHERE title = '{value}'\n",
    "    '''\n",
    "    movie = pd.read_sql_query(myquery, con = engine)\n",
    "    genre_str = ''\n",
    "    for i in range(len(movie)):\n",
    "        genre_str = genre_str+f'{movie.iloc[i,-1]}, '\n",
    "    movie = movie.loc[0]\n",
    "    movie = movie.drop('genre')\n",
    "    movie['genre'] = genre_str\n",
    "    \n",
    "    movie = movie.T.reset_index()\n",
    "    movie = movie.rename({'index':'',0:''},axis = 1)\n",
    "    return ff.create_table(movie)\n",
    "\n",
    "\n",
    "\n",
    "@app.callback(\n",
    "    Output('description', \"children\"),\n",
    "    Input('recommed_dropdown', 'value'),\n",
    ")\n",
    "\n",
    "# call back for the description \n",
    "def display_description(value):\n",
    "    myquery = f'''\n",
    "    SELECT *\n",
    "    FROM movies m\n",
    "    WHERE title = '{value}'\n",
    "    '''\n",
    "    movie = pd.read_sql_query(myquery, con = engine)\n",
    "    return movie['description'][0]\n",
    "\n",
    "@app.callback(\n",
    "    Output('release_map', 'figure'),\n",
    "    Input('recommed_dropdown', 'value'),\n",
    "    )\n",
    "def put_map(value):\n",
    "    myquery = f'''\n",
    "    SELECT  *\n",
    "    FROM movies m\n",
    "    WHERE title = '{value}'\n",
    "    '''\n",
    "    movie = pd.read_sql_query(myquery, con = engine)\n",
    "    id = movie.iloc[0,0]\n",
    "\n",
    "    myquery1 =f'''\n",
    "    SELECT * \n",
    "    FROM movie_release m\n",
    "    INNER JOIN country_code c\n",
    "        ON m.iso_3166_1 = c.iso_3166_1\n",
    "    WHERE m.movie_id = {id}\n",
    "    '''\n",
    "    release = pd.read_sql_query(myquery1, con = engine)\n",
    "    fig = px.choropleth(release, locations=\"alpha3\", hover_name=\"name\", color = \"release_type\", hover_data = ['release_date','release_time'])\n",
    "    return fig\n",
    "\n",
    "\n",
    "\n",
    "@app.callback(\n",
    "    Output('cast_image', 'children'),\n",
    "    Input('recommed_dropdown', 'value'),\n",
    "    State('cast_image','children'),\n",
    ")\n",
    "def cast_image(value, children):\n",
    "    children = []\n",
    "    myquery = f'''\n",
    "    SELECT c.character,\n",
    "        cm.name,\n",
    "        cm.popularity,\n",
    "        cm.profile_path,\n",
    "        cm.known_for_department\n",
    "    FROM movies m\n",
    "    INNER JOIN movie_cast c\n",
    "        ON m.id = c.movie_id\n",
    "    INNER JOIN cast_members cm\n",
    "        ON c.cast_member_id = cm.cast_member_id\n",
    "    WHERE title = '{value}'\n",
    "    ORDER BY cm.popularity DESC\n",
    "    LIMIT 15\n",
    "    '''\n",
    "    cast = pd.read_sql_query(myquery, con = engine)\n",
    "    for i in range(len(cast)):\n",
    "        url = cast['profile_path'][i]\n",
    "        role_name = cast['character'][i]\n",
    "        actor_name = cast['name'][i]\n",
    "        popularity = cast['popularity'][i]\n",
    "        div = html.Div([\n",
    "            html.Div([\n",
    "                 html.Img(src=url, alt='image',style={'height':'150px','width':'100px',\"margin-top\":\"5px\"}),\n",
    "            ],style={ \"display\": \"flex\",\"justify-content\": \"center\"}),\n",
    "            dcc.Markdown(f\"Played: {role_name} \"),\n",
    "            dcc.Markdown(f\"Name: {actor_name}\"),\n",
    "        ],style = {\"display\": \"inline-block\",\"border\":\"2px black solid\",\"margin\":\"5px\"})\n",
    "     \n",
    "        children.append(div)\n",
    "    return children\n",
    "\n",
    "\n",
    "@app.callback(\n",
    "    Output('display_panel', 'children'),\n",
    "    Input('recommed_dropdown', 'value'),\n",
    "    State('display_panel','children'))   \n",
    "def display_comments(value,children):\n",
    "    children = []\n",
    "    myquery = f'''\n",
    "    SELECT  *\n",
    "    FROM movies m\n",
    "    WHERE title = '{value}'\n",
    "    '''\n",
    "    movie = pd.read_sql_query(myquery, con = engine)\n",
    "    id = movie.iloc[0,0]\n",
    "    myquery1 = f'''\n",
    "    SELECT *\n",
    "    FROM reviews\n",
    "    WHERE movie_id = {id}\n",
    "    '''\n",
    "    \n",
    "    reviews = pd.read_sql_query(myquery1, con = engine)\n",
    "    if (len(reviews) == 0):\n",
    "        children.append(html.H2(\"Currently, there is no comment avaiable for this movie.\"))\n",
    "    else:\n",
    "        for i in range(len(reviews)):\n",
    "            div = html.Div([\n",
    "                dcc.Markdown(f'{reviews.iloc[i,4]}'),\n",
    "                html.Div([\n",
    "                    # span for date \n",
    "                    html.Span(f'created on: {reviews.iloc[i,2]}', style = {'margin':\"10px\",\"font-weight\": \"bold\"}),\n",
    "                    html.Span(f'author name: {reviews.iloc[i,1]}',style = {'margin':\"10px\",\"font-weight\": \"bold\"}),\n",
    "                    html.Span(f'rating: {reviews.iloc[i,3]}',style = {'margin':\"10px\",\"font-weight\": \"bold\"})\n",
    "                ])\n",
    "            ], style ={\"border\":\"2px black solid\", 'margin':'10px'})\n",
    "            children.append(div)\n",
    "    return children\n",
    "\n",
    "if __name__== \"__main__\":\n",
    "    app.run(mode= 'external', host='0.0.0.0', port = 8050, debug=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "87cc99e5-6568-462c-9f29-0c97fec3daba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://image.tmdb.org/t/p/original/13AaphwcxenuuWzEx1tAzyg8pA7.jpg'"
      ]
     },
     "execution_count": 354,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cast = pd.read_csv('cast_members.csv')\n",
    "df_cast.iloc[0,5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "a1cf3cc7-709d-4c12-80a9-e1ca606cae85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1075794'"
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "9fb9f9c9-44d3-48c5-95fd-5c5feafffe5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "movie_df.to_csv(\"movies.csv\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "id": "df71aaf2-57b5-4043-8664-23dc4790a7f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "lan_df.to_csv(\"language.csv\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c68e2f0b-2aed-4640-b3cf-b5a7726e99d5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
